packages/components/FocusZone/macos/RCTFocusZone.m

```diff
@@ -0,0 +1,334 @@
+#import "KeyCodes.h"
+#import "RCTFocusZone.h"
+
+typedef enum {
+	FocusZoneActionNone,
+	FocusZoneActionTab,
+	FocusZoneActionShiftTab,
+	FocusZoneActionRightArrow,
+	FocusZoneActionLeftArrow,
+	FocusZoneActionDownArrow,
+	FocusZoneActionUpArrow,
+} FocusZoneAction;
+
+typedef void (^IsViewAtEndOfFocusableRange)(NSView *, BOOL *);
+
+// Maximum vertical (or horizontal) displacement for a candidate view to be
+// enumerated in the same row (or column) as the current focused view
+static const CGFloat FocusZoneBuffer = 3;
+
+@implementation RCTFocusZone
+
+static inline CGFloat GetDistanceBetweenPoints(NSPoint point1, NSPoint point2)
+{
+	NSPoint delta = NSMakePoint(point1.x - point2.x, point1.y - point2.y);
+	return sqrt(delta.x * delta.x + delta.y * delta.y);
+}
+
+static inline NSPoint GetCenterOfRect(NSRect rect)
+{
+	return NSMakePoint(NSMidX(rect), NSMidY(rect));
+}
+
+static inline CGFloat GetDistanceBetweenCentersOfRects(NSRect rect1, NSRect rect2)
+{
+	return GetDistanceBetweenPoints(GetCenterOfRect(rect1), GetCenterOfRect(rect2));
+}
+
+static inline CGFloat GetMinDistanceBetweenRectVerticesAndPoint(NSRect rect, NSPoint point)
+{
+	return fmin(
+		fmin(GetDistanceBetweenPoints(point, NSMakePoint(NSMinX(rect), NSMinY(rect))),
+			GetDistanceBetweenPoints(point, NSMakePoint(NSMaxX(rect), NSMaxY(rect)))),
+		fmin(GetDistanceBetweenPoints(point, NSMakePoint(NSMaxX(rect), NSMinY(rect))),
+			GetDistanceBetweenPoints(point, NSMakePoint(NSMinX(rect), NSMaxY(rect))))
+	);
+}
+
+static NSView *GetFirstResponder(NSWindow *window)
+{
+	NSResponder *responder = [window firstResponder];
+	while (responder != nil && ![responder isKindOfClass:[NSView class]])
+		responder = [responder nextResponder];
+	return [responder isKindOfClass:[NSView class]] ? (NSView *)responder : nil;
+}
+
+static void EnumerateFocusableViews(NSView *root, IsViewAtEndOfFocusableRange block)
```

Comment:
IsViewAtEndOfFocusableRange is this meant to be boolean?

Reply:
See the typedef.  It takes a pointer to BOOL as an output parameter, but returns void.

Reply:
but the callers are directly doing bool * and only this function declaration uses the typedef which seems weird

Reply:
Do you have a recommendation in mind?

For example, should callers define the block before using it so that this type name will appear there?
e.g.
```
IsViewAtEndOfFocusableRange block = ^(NSView *candidateView, BOOL *stop) {...};
EnumerateFocusableViews(self, block);
```

or should the block return a BOOL instead of having an output parameter?

Reply:
right we should acutlaly use the typedef like you just said in the caller